Fix smart pointer usage when dealing with ScriptExecutionContext

chirags27 opened the pull request:
#### 52c77eeabac7a55ff15036c291650bb19d07887e
<pre>
Fix smart pointer usage when dealing with ScriptExecutionContext
<a href="https://bugs.webkit.org/show_bug.cgi?id=254045">https://bugs.webkit.org/show_bug.cgi?id=254045</a>
rdar://106721110

Reviewed by Ryosuke Niwa.

This change fixes uses of ScriptExecutionContext so that it conforms
with the smart pointer usage guidelines.

* Source/WebCore/Modules/speech/SpeechSynthesisUtterance.h:

Canonical link: <a href="https://commits.webkit.org/261817@main">https://commits.webkit.org/261817@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dd4ea19f74adce9222653f086bf0b8a38edf8886

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112941 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/22096 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1616 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4717 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121435 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/117045 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23446 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/13260 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/5906 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118728 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/17412 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100669 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/106031 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/99372 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/1206 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/46436 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14394 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/1246 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/15099 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/10584 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20410 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/53238 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8249 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16950 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/html/ImageBitmap.cpp:
```diff
@@ -110,8 +110,9 @@ RefPtr<ImageBuffer> ImageBitmap::createImageBuffer(ScriptExecutionContext& scrip
+    Ref protectedScriptExecutionContext = scriptExecutionContext;
     WTF::switchOn(source,
-        [&] (auto& specificSource) {
+        [scriptExecutionContext = WTFMove(protectedScriptExecutionContext), &options, &promise] (auto& specificSource) {
```
This can be one liner: scriptExecutionContext = Ref { protectedScriptExecutionContext }

rniwa commented on Source/WebCore/html/ImageBitmap.cpp:
```diff
@@ -145,9 +146,9 @@ void ImageBitmap::createPromise(ScriptExecutionContext& scriptExecutionContext,
+    Ref protectedScriptExecutionContext = scriptExecutionContext;
     WTF::switchOn(source,
-        [&] (auto& specificSource) {
+        [scriptExecutionContext = WTFMove(protectedScriptExecutionContext), &options, left = left, top = top, width = width, height = height, &promise] (auto& specificSource) {
```
Ditto.

rniwa commented on Source/WebCore/css/DOMMatrixReadOnly.cpp:
```diff
@@ -50,9 +50,10 @@ ExceptionOr<Ref<DOMMatrixReadOnly>> DOMMatrixReadOnly::create(ScriptExecutionCon
     return WTF::switchOn(init.value(),
-        [&scriptExecutionContext](const String& init) -> ExceptionOr<Ref<DOMMatrixReadOnly>> {
-            if (!scriptExecutionContext.isDocument())
+        [scriptExecutionContext = WTFMove(protectedScriptExecutionContext)](const String& init) -> ExceptionOr<Ref<DOMMatrixReadOnly>> {
```
This can be one liner: scriptExecutionContext = Ref { scriptExecutionContext }

rniwa commented on Source/WebCore/css/DOMMatrix.cpp:
```diff
@@ -39,9 +39,10 @@ ExceptionOr<Ref<DOMMatrix>> DOMMatrix::create(ScriptExecutionContext& scriptExec
     return WTF::switchOn(init.value(),
-        [&scriptExecutionContext](const String& init) -> ExceptionOr<Ref<DOMMatrix>> {
-            if (!scriptExecutionContext.isDocument())
+        [scriptExecutionContext = WTFMove(protectedScriptExecutionContext)](const String& init) -> ExceptionOr<Ref<DOMMatrix>> {
```
This can be one liner: scriptExecutionContext = Ref { scriptExecutionContext }

rniwa commented on Source/WebCore/bindings/js/JSEventListener.cpp:
```diff
@@ -221,16 +221,19 @@ void JSEventListener::handleEvent(ScriptExecutionContext& scriptExecutionContext
+    Ref protectedScriptExecutionContext = scriptExecutionContext;
+    Ref protectedVM = vm;
+    Ref protectedEvent = event;
+    auto handleExceptionIfNeeded = [scriptExecutionContext = WTFMove(protectedScriptExecutionContext), vm = WTFMove(protectedVM), event = WTFMove(protectedEvent), &jsFunctionGlobalObject] (JSC::Exception* exception) -> bool {
```
Just instantiate Ref inside [~]: [scriptExecutionContext = Ref { scriptExecutionContext }, vm = Ref { vm }, ...]

rniwa commented on Source/WebCore/Modules/speech/SpeechSynthesisUtterance.h:
```diff
@@ -80,14 +80,14 @@ class WEBCORE_EXPORT SpeechSynthesisUtterance final : public PlatformSpeechSynth
     RefPtr<PlatformSpeechSynthesisUtterance> m_platformUtterance;
     RefPtr<SpeechSynthesisVoice> m_voice;
-    ScriptExecutionContext& m_scriptExecutionContext;
+    Ref<ScriptExecutionContext> m_scriptExecutionContext;
```
We probably want to use WeakPtr instead.

rniwa commented on Source/WebCore/bindings/js/JSEventListener.cpp:
```diff
@@ -221,16 +221,16 @@ void JSEventListener::handleEvent(ScriptExecutionContext& scriptExecutionContext
     InspectorInstrumentation::didCallFunction(&scriptExecutionContext);
 
-    auto handleExceptionIfNeeded = [&] (JSC::Exception* exception) -> bool {
+    auto handleExceptionIfNeeded = [scriptExecutionContext = Ref { scriptExecutionContext }, vm = Ref { vm }, event = Ref { event }, &jsFunctionGlobalObject] (JSC::Exception* exception) -> bool {
```
Oh, this lambda is never stored anywhere so no need to make this change.

rniwa commented on Source/WebCore/css/DOMMatrix.cpp:
```diff
@@ -40,8 +40,8 @@ ExceptionOr<Ref<DOMMatrix>> DOMMatrix::create(ScriptExecutionContext& scriptExec
     return WTF::switchOn(init.value(),
-        [&scriptExecutionContext](const String& init) -> ExceptionOr<Ref<DOMMatrix>> {
-            if (!scriptExecutionContext.isDocument())
+        [scriptExecutionContext = Ref { scriptExecutionContext }](const String& init) -> ExceptionOr<Ref<DOMMatrix>> {
```
Same thing here. This lambda is immediately executed so no need to create a new Ref.

rniwa commented on Source/WebCore/css/DOMMatrixReadOnly.cpp:
```diff
@@ -51,8 +51,8 @@ ExceptionOr<Ref<DOMMatrixReadOnly>> DOMMatrixReadOnly::create(ScriptExecutionCon
     return WTF::switchOn(init.value(),
-        [&scriptExecutionContext](const String& init) -> ExceptionOr<Ref<DOMMatrixReadOnly>> {
-            if (!scriptExecutionContext.isDocument())
+        [scriptExecutionContext = Ref { scriptExecutionContext }](const String& init) -> ExceptionOr<Ref<DOMMatrixReadOnly>> {
```
Ditto.

rniwa commented on Source/WebCore/html/ImageBitmap.cpp:
```diff
@@ -110,8 +110,9 @@ RefPtr<ImageBuffer> ImageBitmap::createImageBuffer(ScriptExecutionContext& scrip
+    Ref protectedScriptExecutionContext = scriptExecutionContext;
     WTF::switchOn(source,
-        [&] (auto& specificSource) {
+        [scriptExecutionContext = Ref { scriptExecutionContext }, &options, &promise] (auto& specificSource) {
```
Ditto.

rniwa commented on Source/WebCore/html/ImageBitmap.cpp:
```diff
@@ -145,9 +146,9 @@ void ImageBitmap::createPromise(ScriptExecutionContext& scriptExecutionContext,
+    Ref protectedScriptExecutionContext = scriptExecutionContext;
     WTF::switchOn(source,
-        [&] (auto& specificSource) {
+        [scriptExecutionContext = Ref { scriptExecutionContext }, &options, left = left, top = top, width = width, height = height, &promise] (auto& specificSource) {
```
Ditto.

chirags27 committed:
Fix smart pointer usage when dealing with ScriptExecutionContext
https://bugs.webkit.org/show_bug.cgi?id=254045
rdar://106721110

Reviewed by Ryosuke Niwa.

This change fixes uses of ScriptExecutionContext so that it conforms
with the smart pointer usage guidelines.

* Source/WebCore/Modules/speech/SpeechSynthesisUtterance.h:

Canonical link: https://commits.webkit.org/261817@main
